Feature save scraped events to db

twumm opened the pull request:
# Description
Create feature to scrape events from eventbrite

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Change status
- [x] Complete, tested, ready to review and merge
- [ ] Complete, but not tested (may need new tests)
- [ ] Incomplete/work-in-progress, PR is for discussion/feedback

# How Has This Been Tested?
We used Postman to make calls to the api route. Data from eventbrite got saved into the database

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation, if relevant
- [] I have run the app using my feature and ensured that no functionality is broken
- [x] My changes generate no new warnings
- [x] There are no merge conflicts

twumm committed:
feat: create event model and controller

Add schema for the event data type
Add method to pass scraped events and save them in the database
Install mongoose-unique-validator

twumm committed:
Merge branch 'develop' into feature-save-scraped-events-to-db

twumm commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Create endpoint to save data from web-scraper](https://trello.com/c/SFC6RYon/47-create-endpoint-to-save-data-from-web-scraper)

twumm committed:
feat: create method to fetch and save events

twumm committed:
feat: refactor route and controller for fetching events from the scraper

Save events into the database
Send events fetched from the scraper as a response

twumm committed:
refactor: return promise for the webScraper module

abidex4yemi commented on resources/event/event.controller.js:
```diff
@@ -0,0 +1,29 @@
+const ScrapeEvents = require("../../utils/webScraper");
+
+const addScrapedEvent = async () => {
+  const scrappedEventsArray = await ScrapeEvents;
```
Could you elaborate on this line, I think `ScrapeEvents` it should be invoked as a function.

twumm replied:
This is a Promise. We tried invoking earlier but it did not work.
We will be happy to get some insight on this.

abidex4yemi commented on resources/event/event.controller.js:
```diff
@@ -0,0 +1,29 @@
+
+const addScrapedEvent = async () => {
+  const scrappedEventsArray = await ScrapeEvents;
+  await scrappedEventsArray.forEach(async (event) => {
```
Instead of looping through the event and inserting one after each, this isn't efficient enough instead let's use bulk insert without looping through the event array for better perfomance

twumm replied:
I agree. We will refactor to use `insertMany()`.

abidex4yemi commented on resources/event/event.controller.js:
```diff
@@ -0,0 +1,29 @@
+  return scrappedEventsArray;
+};
+
+const findEvent = async () => {
+  const events = await Event.find();
+  return events;
+};
```
This code here could be optimized and move to a single function

twumm replied:
@abidex4yemi This is a single function.

Can you elaborate further? Do you mean handling error cases etc?

abidex4yemi commented on utils/webScraper.js:
```diff
@@ -1,89 +1,85 @@
+const fetchData = async () => {
+  const result = await axios({
+    method: "get",
+    url: `https://www.eventbrite.com/d/nigeria--${locationSample}/${science}--events/`,
```
This `URL` should be more dynamic and location should not be hardcoded

twumm replied:
Yes. This will be refactored. 
We have created a card for making the scraping dynamic(https://trello.com/c/FOeY3RSV/48-make-scaping-more-dynamic). That will be our focus for tomorrow.

This is static for now.

abidex4yemi commented on utils/webScraper.js:
```diff
@@ -1,89 +1,85 @@
     const newTitles = titlesArray.filter((str) => /\S/.test(str));
-
-    for (let i = 0; i < datesArray.length; i++) {
+    const uniqueTitles = Array.from(new Set(newTitles));
+    const uniqueDates = Array.from(new Set(datesArray));
+    const uniqueLinkIds = Array.from(new Set(linksArray));
+    const uniqueImageLinks = Array.from(new Set(imagesArray));
+    const uniqueLocations = Array.from(new Set(locationsArray));
+    const uniquePrices = Array.from(new Set(pricesArray));
+    const uniqueEventLinks = Array.from(new Set(eventLinksArray));
+    for (let i = 0; i < uniqueDates.length; i++) {
```
This could be cleaner and D>R>Y than this

twumm replied:
We can work on this in the next iteration or maybe tomorrow when we attempt the dynamic scraper.

abidex4yemi reviewed (changes requested):
Good job on this PR but I think we can make this code more efficient and cleaner

twumm committed:
refactor: use mongoose's insertMany method to insert events

Update model schema to make it possible to receive other events
Update webScraper array object keys to match schema in the db

abidex4yemi reviewed (approved):
Looks good.
